File: benchmarks/micro/AggregatePhiFieldForwarding.c
#include <stdint.h>
#include <stdbool.h>
#include <unistd.h>

typedef struct Pair {
    int64_t value;
    int64_t tag;
} Pair;

static int64_t read_value(bool flag, int64_t value, int64_t tag) {
    Pair pair = flag
        ? (Pair){ value, tag }
        : (Pair){ value, tag + 1 };
    return pair.value;
}

int main(void) {
    int64_t total = (int64_t)getpid();

    for (int32_t i = 0; i < 300000; i += 1) {
        total += read_value((i & 1) == 0, total % 1000000007, (int64_t)i);
        total %= 1000000007;
    }

    return total <= 0 ? 1 : 0;
}

File: benchmarks/micro/AggregatePhiFieldForwarding.rs
#[allow(dead_code)]
struct Pair {
    value: i64,
    tag: i64,
}

fn read_value(flag: bool, value: i64, tag: i64) -> i64 {
    let pair = if flag {
        Pair { value, tag }
    } else {
        Pair {
            value,
            tag: tag + 1,
        }
    };
    pair.value
}

fn main() {
    let mut total = std::process::id() as i64;

    for i in 0_i32..300_000 {
        total += read_value((i & 1) == 0, total % 1_000_000_007, i as i64);
        total %= 1_000_000_007;
    }

    if total <= 0 {
        std::process::exit(1);
    }
}
